lib/elixir/lib/supervisor.ex

```diff
@@ -518,26 +518,58 @@ defmodule Supervisor do
           | {:max_restarts, non_neg_integer}
           | {:max_seconds, pos_integer}
 
+  @typedoc "Supported restart options"
+  @type restart :: :permanent | :transient | :temporary
+
+  # TODO: Update :shutdown to "timeout() | :brutal_kill" when we require Erlang/OTP 24.
+  #       Additionally apply https://github.com/elixir-lang/elixir/pull/11836
+  @typedoc "Supported shutdown options"
+  @type shutdown :: pos_integer() | :infinity | :brutal_kill
+
   @typedoc "Supported strategies"
   @type strategy :: :one_for_one | :one_for_all | :rest_for_one
 
+  @typedoc """
+  Supervisor type
+
+  Whether the supervisor is a worker or a supervisor.
+  """
+  @type type :: :worker | :supervisor
+
   # Note we have inlined all types for readability
-  @typedoc "The supervisor specification"
+  @typedoc """
+  The supervisor child specification.
+
+  It defines how the supervisor should start, stop and restart each of its children.
+  """
   @type child_spec :: %{
           required(:id) => atom() | term(),
-          required(:start) => {module(), function_name :: atom(), args :: [term()]},
-          optional(:restart) => :permanent | :transient | :temporary,
-          # TODO: Update :shutdown to "timeout() | :brutal_kill" when we require Erlang/OTP 24.
-          #       Additionally apply https://github.com/elixir-lang/elixir/pull/11836
-          optional(:shutdown) => pos_integer() | :infinity | :brutal_kill,
-          optional(:type) => :worker | :supervisor,
+          required(:start) => {module(), function_name :: atom(), args :: [term()] | :undefined},
```

Comment:
:undefined added.
It is used internally, but still part of `:supervisor.child_spec()`

Reply:
I would not mix those. Passing `:undefined` is never a valid input. If it s a valid output, then we should create a separate type for that on the API that uses it.